protocols/gossipsub/src/metrics.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright 2020 Sigma Prime Pty Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A set of metrics used to help track and diagnose the network behaviour of the gossipsub
+//! protocol.
+
+use std::collections::HashMap;
+
+use open_metrics_client::encoding::text::Encode;
+use open_metrics_client::metrics::counter::Counter;
+use open_metrics_client::metrics::family::Family;
+use open_metrics_client::metrics::gauge::Gauge;
+use open_metrics_client::registry::Registry;
+
+use crate::topic::TopicHash;
+
+// Default value that limits how many topics for which there has been a previous or current
+// subscription do we store metrics.
+const DEFAULT_MAX_EXPLICITLY_SUBSCRIBED_TOPICS: usize = 300;
```

Comment:
While a bit of a hack in my eyes, this seems like a practical solution to the problem of metric churn.

What is your experience with this constant limit? Do you ever hit it on your production nodes?

Reply:
For us the subscriptions that are really important occur at the beginning and are few, this is a very high upper limit. Still, that's why I left it as an application-configurable value

Reply:
We're currently bounded at about 250 topics (we can double our topic subscription around a fork boundary), most operating nodes would have around 125. 
I guess we've set this value selfishly around our needs a little, but hopefully is sufficiently large for others.